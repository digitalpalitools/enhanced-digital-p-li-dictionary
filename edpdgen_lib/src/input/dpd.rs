// NOTE: Keep the order deliberately randomized as we support column reordering.
#[derive(Debug, Serialize, Deserialize)]
pub struct DpdPaliWord {
    #[serde(rename = "Pāli1")]
    pub pali1: String,
    #[serde(rename = "Pāli2")]
    pub pali2: String,
    #[serde(rename = "Fin")]
    pub fin: String,
    #[serde(rename = "POS")]
    pub pos: String,
    #[serde(rename = "Grammar")]
    pub grammar: String,
    #[serde(rename = "Derived from")]
    pub derived_from: String,
    #[serde(rename = "Neg")]
    pub neg: String,
    #[serde(rename = "Verb")]
    pub verb: String,
    #[serde(rename = "Trans")]
    pub trans: String,
    #[serde(rename = "Case")]
    pub case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    pub in_english: String,
    #[serde(rename = "Sanskrit")]
    pub sanskrit: String,
    #[serde(rename = "Sk Root")]
    pub sanskrit_root: String,
    #[serde(rename = "Family")]
    pub family: String,
    #[serde(rename = "Pāli Root")]
    pub pali_root: String,
    #[serde(rename = "V")]
    pub v: String,
    #[serde(rename = "Grp")]
    pub grp: String,
    #[serde(rename = "Sgn")]
    pub sgn: String,
    #[serde(rename = "Root Meaning")]
    pub root_meaning: String,
    #[serde(rename = "Base")]
    pub base: String,
    #[serde(rename = "Construction")]
    pub construction: String,
    #[serde(rename = "Derivative")]
    pub derivative: String,
    #[serde(rename = "Suffix")]
    pub suffix: String,
    #[serde(rename = "Compound")]
    pub compound: String,
    #[serde(rename = "Compound Construction")]
    pub compound_construction: String,
    #[serde(rename = "Source1")]
    pub source1: String,
    #[serde(rename = "Sutta1")]
    pub sutta1: String,
    #[serde(rename = "Example1")]
    pub example1: String,
    #[serde(rename = "Source 2")]
    pub source2: String,
    #[serde(rename = "Sutta2")]
    pub sutta2: String,
    #[serde(rename = "Example 2")]
    pub example2: String,
    #[serde(rename = "Antonyms")]
    pub antonyms: String,
    #[serde(rename = "Synonyms – different word")]
    pub synonyms: String,
    #[serde(rename = "Variant – same constr or diff reading")]
    pub variant: String,
    #[serde(rename = "Commentary")]
    pub commentary: String,
    #[serde(rename = "Literal Meaning")]
    pub literal_meaning: String,
    #[serde(rename = "Root In Comps")]
    pub root_in_compound: String,
    #[serde(rename = "Notes")]
    pub notes: String,
    #[serde(rename = "Stem")]
    pub stem: String,
    #[serde(rename = "Pattern")]
    pub pattern: String,
    #[serde(rename = "Buddhadatta")]
    pub buddhadatta: String,
    #[serde(rename = "Cl")]
    pub cl: String,
}
