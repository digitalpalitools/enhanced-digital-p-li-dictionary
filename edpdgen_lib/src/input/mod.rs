pub mod dpd;
pub mod dps;
pub mod input_format;
use csv::Reader;
use pls_core_extras::logger::PlsLogger;
use regex::{Captures, Regex};
use std::fs::File;
use std::path::Path;

lazy_static! {
    pub static ref PALI1_CRACKER: Regex =
        Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

pub fn create_csv_reader(path: &Path, logger: &dyn PlsLogger) -> Result<Reader<File>, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    Ok(rdr)
}

pub fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_case::test_case;

    #[test_case("ābādha", "ābādha"; "0 digits")]
    #[test_case("adhikāra 1", "adhikāra 001"; "1 digit")]
    #[test_case("adhikāra 10", "adhikāra 010"; "2 digits")]
    fn test_sort_key(id: &str, expected_sk: &str) {
        let sk = make_sort_key(id);

        assert_eq!(sk, expected_sk)
    }
}
