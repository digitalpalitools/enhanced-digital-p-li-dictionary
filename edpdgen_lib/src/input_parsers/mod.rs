use crate::EdpdLogger;
use regex::{Captures, Regex};
use std::path::Path;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "toc_summary",
            include_str!("templates/toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "word_data",
            include_str!("templates/word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

pub trait PaliWord {
    fn sort_key(&self) -> String;
    fn group_id(&self) -> String;
    fn toc_id(&self) -> String;
    fn toc_entry(&self) -> Result<String, String>;
    fn word_data_entry(&self, short_name: &str) -> Result<String, String>;
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DpdPaliWord {
    #[serde(rename = "Pāli1")]
    pali1: String,
    #[serde(rename = "Pāli2")]
    pali2: String,
    #[serde(rename = "Fin")]
    fin: String,
    #[serde(rename = "POS")]
    pos: String,
    #[serde(rename = "Grammar")]
    grammar: String,
    #[serde(rename = "Derived from")]
    derived_from: String,
    #[serde(rename = "Neg")]
    neg: String,
    #[serde(rename = "Verb")]
    verb: String,
    #[serde(rename = "Trans")]
    trans: String,
    #[serde(rename = "Case")]
    case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    in_english: String,
    #[serde(rename = "Sanskrit")]
    sanskrit: String,
    #[serde(rename = "Sk Root")]
    sanskrit_root: String,
    #[serde(rename = "Family")]
    family: String,
    #[serde(rename = "Pāli Root")]
    pali_root: String,
    #[serde(rename = "V")]
    v: String,
    #[serde(rename = "Grp")]
    grp: String,
    #[serde(rename = "Sgn")]
    sgn: String,
    #[serde(rename = "Root Meaning")]
    root_meaning: String,
    #[serde(rename = "Base")]
    base: String,
    #[serde(rename = "Construction")]
    construction: String,
    #[serde(rename = "Derivative")]
    derivative: String,
    #[serde(rename = "Suffix")]
    suffix: String,
    #[serde(rename = "Compound")]
    compound: String,
    #[serde(rename = "Compound Construction")]
    compound_construction: String,
    #[serde(rename = "Source1")]
    source1: String,
    #[serde(rename = "Sutta1")]
    sutta1: String,
    #[serde(rename = "Example1")]
    example1: String,
    #[serde(rename = "Source 2")]
    source2: String,
    #[serde(rename = "Sutta2")]
    sutta2: String,
    #[serde(rename = "Example 2")]
    example2: String,
    #[serde(rename = "Antonyms")]
    antonyms: String,
    #[serde(rename = "Synonyms – different word")]
    synonyms: String,
    #[serde(rename = "Variant – same constr or diff reading")]
    variant: String,
    #[serde(rename = "Commentary")]
    commentary: String,
    #[serde(rename = "Notes")]
    notes: String,
    #[serde(rename = "Stem")]
    stem: String,
    #[serde(rename = "Pattern")]
    pattern: String,
    #[serde(rename = "Buddhadatta")]
    buddhadatta: String,
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
    toc_id: &'a str,
    short_name: &'a str,
}

impl PaliWord for DpdPaliWord {
    fn sort_key(&self) -> String {
        let sk = PALI1_CRACKER.replace(&self.pali1, |caps: &Captures| {
            // NOTE: Best case effort. Not sweating it.
            let n = &caps[3].parse::<i32>().unwrap_or(0);
            format!("{} {:03}", &caps[1], n)
        });

        sk.into_owned()
    }

    fn group_id(&self) -> String {
        let gid = PALI1_CRACKER.replace(&self.pali1, |caps: &Captures| caps[1].to_string());

        gid.into_owned()
    }

    fn toc_id(&self) -> String {
        self.pali1.replace(" ", "_") + "_dpd"
    }

    fn toc_entry(&self) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("toc_id", &self.toc_id());
        context.insert("pali1", &self.pali1);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);

        TEMPLATES
            .render("toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(&self, short_name: &str) -> Result<String, String> {
        let vm = WordDataViewModel {
            word: &self,
            toc_id: &self.toc_id(),
            short_name,
        };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("word_data", &context)
            .map_err(|e| e.to_string())
    }
}

pub fn load_words<'a>(
    path: &Path,
    logger: &'a impl EdpdLogger,
) -> Result<impl Iterator<Item = impl PaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<DpdPaliWord>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "ābādha"; "0 digits")]
    #[test_case(9, "adhikāra 001"; "1 digit")]
    #[test_case(11, "adhikāra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| r.sort_key())
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }

    #[test_case(4, "abahulīkata"; "0 digits")]
    #[test_case(5, "abala"; "1 digit")]
    #[test_case(11, "adhikāra"; "2 digits")]
    fn test_group_id(rec_number: usize, expected_gid: &str) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("failed to load");

        let gid = recs
            .nth(rec_number)
            .map(|r| r.group_id())
            .expect("unexpected");

        assert_eq!(gid, expected_gid);
    }

    #[test_case(4, "abahulīkata_dpd"; "0 digits")]
    #[test_case(5, "abala_1_dpd"; "1 digit")]
    #[test_case(11, "adhikāra_10_dpd"; "2 digits")]
    fn test_toc_id(rec_number: usize, expected_toc_id: &str) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("unexpected");

        let toc_id = recs
            .nth(rec_number)
            .map(|r| r.toc_id())
            .expect("unexpected");

        assert_eq!(toc_id, expected_toc_id);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn toc_summary_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry().expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words(&get_csv_path(), &l).expect("unexpected");

        let word_data = recs
            .nth(rec_number)
            .map(|r| r.word_data_entry("dpd").expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
