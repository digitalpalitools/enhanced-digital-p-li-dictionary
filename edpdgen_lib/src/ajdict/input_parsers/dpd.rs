use crate::ajdict::AjDictPaliWord;
use crate::input::dpd::DpdPaliWord;
use crate::input::make_sort_key;
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dpd_word_data",
            include_str!("templates/dpd_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera
    };
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
}

impl AjDictPaliWord for DpdPaliWord {
    fn id(&self) -> &str {
        &self.pali1
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn word_data_entry(&self) -> Result<String, String> {
        let vm = WordDataViewModel { word: &self };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dpd_word_data", &context)
            .map(|x| x.replace("\r", "").replace("\n", ""))
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ajdict::input_parsers::load_words;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let word_data = recs
            .nth(rec_number)
            .map(|r| r.word_data_entry().expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
