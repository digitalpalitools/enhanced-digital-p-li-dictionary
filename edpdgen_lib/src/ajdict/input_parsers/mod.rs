use crate::ajdict::AjDictPaliWord;
use crate::EdpdLogger;
use regex::{Captures, Regex};
use std::path::Path;

pub mod dpd;

lazy_static! {
    static ref PALI1_CRACKER: Regex = Regex::new(r"(.*)( )(\d+)$").expect("Malformed regex string");
}

//
//
//
//
//  DUP code
//
//
//
//
//
//
pub fn load_words<'a, T: 'a + serde::de::DeserializeOwned + AjDictPaliWord>(
    path: &Path,
    logger: &'a dyn EdpdLogger,
) -> Result<impl Iterator<Item = impl AjDictPaliWord> + 'a, String> {
    logger.info(&format!("Loading words from {:?}.", path));

    let file = std::fs::File::open(path).map_err(|e| e.to_string())?;
    let rdr = csv::ReaderBuilder::new()
        .delimiter(b'\t')
        .flexible(true)
        .from_reader(file);

    let words = rdr
        .into_deserialize::<T>()
        .enumerate()
        .filter_map(move |(i, r)| match r {
            Ok(w) => Some(w),
            Err(e) => {
                logger.error(&format!(
                    "Unable to deserialize record #{}. Error: {}.",
                    i, e
                ));
                None
            }
        });

    logger.info(&format!("... done loading words from {:?}.", &path));
    Ok(words)
}

//
//
//
//
//  DUP code
//
//
//
//
//
//
fn make_sort_key(id: &str) -> String {
    let sk = PALI1_CRACKER.replace(id, |caps: &Captures| {
        // NOTE: Best case effort. Not sweating it.
        let n = &caps[3].parse::<i32>().unwrap_or(0);
        format!("{} {:03}", &caps[1], n)
    });

    sk.into_owned()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::input::dpd::DpdPaliWord;
    use crate::resolve_file_in_manifest_dir;
    use crate::tests::TestLogger;
    use std::path::PathBuf;
    use test_case::test_case;

    pub fn get_csv_path() -> PathBuf {
        resolve_file_in_manifest_dir("Pali_English_Dictionary_10_rows-full.csv")
            .expect("must exist!")
    }

    #[test_case(0, "훮b훮dha"; "0 digits")]
    #[test_case(9, "adhik훮ra 001"; "1 digit")]
    #[test_case(11, "adhik훮ra 010"; "2 digits")]
    fn test_sort_key(rec_number: usize, expected_sk: &str) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("failed to load");

        let sk = recs
            .nth(rec_number)
            .map(|r| make_sort_key(r.id()))
            .expect("unexpected");

        assert_eq!(sk, expected_sk)
    }
}
