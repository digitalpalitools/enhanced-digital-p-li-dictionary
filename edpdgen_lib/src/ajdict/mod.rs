use crate::inflection_generator::InflectionGenerator;
use crate::input::dpd::DpdPaliWord;
use crate::input::input_format::InputFormat;
use crate::{DictionaryBuilder, DictionaryFile, DictionaryInfo, EdpdLogger};
use std::path::Path;

mod input_parsers;
mod output_generators;

pub struct AjDict<'a> {
    dict_info: &'a DictionaryInfo<'a>,
    input_data_path: &'a Path,
    logger: &'a dyn EdpdLogger,
}

impl<'a> DictionaryBuilder<'a> for AjDict<'a> {
    fn new(
        dict_info: &'a DictionaryInfo,
        input_data_path: &'a Path,
        _igen: &'a dyn InflectionGenerator,
        logger: &'a dyn EdpdLogger,
    ) -> Self {
        AjDict {
            dict_info,
            input_data_path,
            logger,
        }
    }

    fn build_files(&self) -> Result<Vec<DictionaryFile>, String> {
        match self.dict_info.input_format {
            InputFormat::Dpd => run_for_ods_type::<DpdPaliWord>(self.input_data_path, self.logger),
            InputFormat::Dps => todo!(),
        }
    }
}

pub trait AjDictPaliWord {
    fn id(&self) -> &str;
    fn sort_key(&self) -> String;
    fn word_data_entry(&self) -> Result<String, String>;
}

pub fn run_for_ods_type<'a, T: 'a + serde::de::DeserializeOwned + AjDictPaliWord>(
    input_data_path: &Path,
    logger: &dyn EdpdLogger,
) -> Result<Vec<DictionaryFile>, String> {
    let words = input_parsers::load_words::<T>(input_data_path, logger)?;
    let sd_files = output_generators::create_dictionary(words, logger)?;

    Ok(sd_files)
}
