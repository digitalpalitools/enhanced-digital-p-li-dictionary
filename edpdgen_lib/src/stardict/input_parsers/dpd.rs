use crate::inflection_generator::InflectionGenerator;
use crate::stardict::input_parsers::{make_group_id, make_sort_key, make_toc_id, PaliWord};
use tera::{Context, Tera};

lazy_static! {
    static ref TEMPLATES: Tera = {
        let mut tera = Tera::default();
        tera.add_raw_templates(vec![(
            "dpd_toc_summary",
            include_str!("templates/dpd_toc_summary.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.add_raw_templates(vec![(
            "dpd_word_data",
            include_str!("templates/dpd_word_data.html"),
        )])
        .expect("Unexpected failure adding template");
        tera.autoescape_on(vec!["html"]);
        tera
    };
}

// NOTE: Keep the order deliberately randomized as we support column reordering.
#[derive(Debug, Serialize, Deserialize)]
pub struct DpdPaliWord {
    #[serde(rename = "Pāli1")]
    pali1: String,
    #[serde(rename = "Pāli2")]
    pali2: String,
    #[serde(rename = "Fin")]
    fin: String,
    #[serde(rename = "POS")]
    pos: String,
    #[serde(rename = "Grammar")]
    grammar: String,
    #[serde(rename = "Derived from")]
    derived_from: String,
    #[serde(rename = "Neg")]
    neg: String,
    #[serde(rename = "Verb")]
    verb: String,
    #[serde(rename = "Trans")]
    trans: String,
    #[serde(rename = "Case")]
    case: String,
    #[serde(rename = "Meaning IN CONTEXT")]
    in_english: String,
    #[serde(rename = "Sanskrit")]
    sanskrit: String,
    #[serde(rename = "Sk Root")]
    sanskrit_root: String,
    #[serde(rename = "Family")]
    family: String,
    #[serde(rename = "Pāli Root")]
    pali_root: String,
    #[serde(rename = "V")]
    v: String,
    #[serde(rename = "Grp")]
    grp: String,
    #[serde(rename = "Sgn")]
    sgn: String,
    #[serde(rename = "Root Meaning")]
    root_meaning: String,
    #[serde(rename = "Base")]
    base: String,
    #[serde(rename = "Construction")]
    construction: String,
    #[serde(rename = "Derivative")]
    derivative: String,
    #[serde(rename = "Suffix")]
    suffix: String,
    #[serde(rename = "Compound")]
    compound: String,
    #[serde(rename = "Compound Construction")]
    compound_construction: String,
    #[serde(rename = "Source1")]
    source1: String,
    #[serde(rename = "Sutta1")]
    sutta1: String,
    #[serde(rename = "Example1")]
    example1: String,
    #[serde(rename = "Source 2")]
    source2: String,
    #[serde(rename = "Sutta2")]
    sutta2: String,
    #[serde(rename = "Example 2")]
    example2: String,
    #[serde(rename = "Antonyms")]
    antonyms: String,
    #[serde(rename = "Synonyms – different word")]
    synonyms: String,
    #[serde(rename = "Variant – same constr or diff reading")]
    variant: String,
    #[serde(rename = "Commentary")]
    commentary: String,
    #[serde(rename = "Literal Meaning")]
    literal_meaning: String,
    #[serde(rename = "Root In Comps")]
    root_in_compound: String,
    #[serde(rename = "Notes")]
    notes: String,
    #[serde(rename = "Stem")]
    stem: String,
    #[serde(rename = "Pattern")]
    pattern: String,
    #[serde(rename = "Buddhadatta")]
    buddhadatta: String,
    #[serde(rename = "Cl")]
    cl: String,
}

#[derive(Serialize)]
struct WordDataViewModel<'a> {
    word: &'a DpdPaliWord,
    toc_id: &'a str,
    short_name: &'a str,
    feedback_form_url: &'a str,
    host_url: &'a str,
    host_version: &'a str,
    inflection_table: &'a str,
}

impl PaliWord for DpdPaliWord {
    fn id(&self) -> &str {
        &self.pali1
    }

    fn sort_key(&self) -> String {
        make_sort_key(&self.id())
    }

    fn group_id(&self) -> String {
        make_group_id(&self.id())
    }

    fn toc_id(&self, short_name: &str) -> String {
        make_toc_id(&self.id(), short_name)
    }

    fn toc_entry(&self, short_name: &str) -> Result<String, String> {
        let mut context = Context::new();
        context.insert("short_name", short_name);
        context.insert("toc_id", &self.toc_id(short_name));
        context.insert("pali1", &self.pali1);
        context.insert("pos", &self.pos);
        context.insert("in_english", &self.in_english);
        context.insert("buddhadatta", &self.buddhadatta);

        TEMPLATES
            .render("dpd_toc_summary", &context)
            .map_err(|e| e.to_string())
    }

    fn word_data_entry(
        &self,
        short_name: &str,
        feedback_form_url: &str,
        host_url: &str,
        host_version: &str,
        igen: &dyn InflectionGenerator,
    ) -> Result<String, String> {
        let vm = WordDataViewModel {
            word: &self,
            toc_id: &self.toc_id(short_name),
            short_name,
            feedback_form_url,
            host_url,
            host_version,
            inflection_table: &igen.generate_inflection_table_html(&self.pali1),
        };

        let context = Context::from_serialize(&vm).map_err(|e| e.to_string())?;
        TEMPLATES
            .render("dpd_word_data", &context)
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::stardict::input_parsers::load_words;
    use crate::stardict::input_parsers::tests::get_csv_path;
    use crate::tests::{TestInflectionGenerator, TestLogger};
    use test_case::test_case;

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn toc_summary_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");

        let toc_summary = recs
            .nth(rec_number)
            .map(|r| r.toc_entry("dpy").expect("unexpected"))
            .expect("unexpected");

        insta::assert_snapshot!(toc_summary);
    }

    #[test_case(0)]
    #[test_case(1)]
    #[test_case(2)]
    #[test_case(3)]
    #[test_case(4)]
    #[test_case(5)]
    #[test_case(6)]
    #[test_case(7)]
    #[test_case(8)]
    #[test_case(9)]
    #[test_case(10)]
    #[test_case(11)]
    #[test_case(12)]
    fn word_data_tests(rec_number: usize) {
        let l = TestLogger::new();
        let mut recs = load_words::<DpdPaliWord>(&get_csv_path(), &l).expect("unexpected");
        let igen = TestInflectionGenerator::new();

        let word_data = recs
            .nth(rec_number)
            .map(|r| {
                r.word_data_entry("dpd", "fb_url", "host url", "host version", &igen)
                    .expect("unexpected")
            })
            .expect("unexpected");

        insta::assert_snapshot!(word_data);
    }
}
