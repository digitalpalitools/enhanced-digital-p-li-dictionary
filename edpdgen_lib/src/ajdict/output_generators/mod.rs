use crate::ajdict::AjDictPaliWord;
use crate::{DictionaryFile, EdpdLogger};

fn create_dict_entries(
    words: impl Iterator<Item = impl AjDictPaliWord>,
    logger: &dyn EdpdLogger,
) -> Result<(Vec<String>, Vec<String>), String> {
    logger.info(&"Creating dict entries.".to_string());

    let mut dpd_entries: Vec<String> = Vec::new();
    let mut cdpd_entries: Vec<String> = Vec::new();
    for (n, word) in words.into_iter().enumerate() {
        dpd_entries.push(word.word_data_entry()?);
        cdpd_entries.push(word.concise_word_data_entry()?);

        if n % 1_000 == 0 && n != 0 {
            logger.info(&format!(
                "... created {:05} dict entries, ending with '{}'.",
                n,
                word.id()
            ));
        }
    }

    logger.info(&format!(
        "... done creating {} dict entries.",
        dpd_entries.len()
    ));

    Ok((dpd_entries, cdpd_entries))
}

fn create_txt_data(dict_entries: Vec<String>, logger: &dyn EdpdLogger) -> Vec<u8> {
    logger.info(&format!(
        "Creating dict data for {} txt entries.",
        &dict_entries.len()
    ));

    let txt: Vec<u8> = dict_entries.iter().fold(Vec::new(), |mut acc, e| {
        acc.append(&mut e.to_owned().into_bytes());
        acc.push(0x0d);
        acc.push(0x0a);
        acc
    });

    logger.info(&format!(
        "... done creating {} idx entries.",
        &dict_entries.len()
    ));

    txt
}

pub fn create_dictionary(
    words: impl Iterator<Item = impl AjDictPaliWord>,
    logger: &dyn EdpdLogger,
) -> Result<Vec<DictionaryFile>, String> {
    let (dpd_entries, cdpd_entries) = create_dict_entries(words, logger)?;
    let dpd_txt = create_txt_data(dpd_entries, logger);
    let cdpd_txt = create_txt_data(cdpd_entries, logger);

    Ok(vec![
        DictionaryFile {
            extension: "dpd.ajd.txt".to_string(),
            bom: vec![0xEF, 0xBB, 0xBF],
            data: dpd_txt,
            can_be_empty: false,
        },
        DictionaryFile {
            extension: "cdpd.ajd.txt".to_string(),
            bom: vec![0xEF, 0xBB, 0xBF],
            data: cdpd_txt,
            can_be_empty: false,
        },
    ])
}

#[cfg(test)]
mod tests {
    use crate::ajdict::output_generators::create_txt_data;
    use crate::tests::TestLogger;

    #[test]
    fn create_dict_test() {
        let dict_entries = vec!["a;1;2".to_string(), "b;3;4".to_string()];

        let txt_data = create_txt_data(dict_entries, &TestLogger::new());

        assert_eq!(
            txt_data,
            vec![97, 59, 49, 59, 50, 13, 10, 98, 59, 51, 59, 52, 13, 10]
        );
    }
}
