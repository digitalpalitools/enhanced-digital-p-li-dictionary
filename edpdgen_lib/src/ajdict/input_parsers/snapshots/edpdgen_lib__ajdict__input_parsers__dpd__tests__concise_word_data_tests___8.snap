---
source: edpdgen_lib/src/ajdict/input_parsers/dpd.rs
expression: word_data

---
ābandhaka; adj; tying, connecting, fixing
